src/transformers/generation/utils.py

```diff
@@ -3404,184 +3491,304 @@ def _beam_search(
+            1. Each item in batch has `num_beams` * `vocab_size` candidate continuations. For each item, get the
+               top K [K = (number of EOS tokens + 1) * `num_beams`] candidates with the highest accumulated
+               log-probabilities, or sample them without replacement using the accumulated scores
+            2. We gather the top K (as opposed to `num_beams`, or any number lower than K) here so that we have at
+               least `num_beams` sequences remaining to continue the live beam search.
+            3. Note that other stopping criteria might result in impossible to continue beams, i.e. all continuations
+               selected in this step hit the stopping criteria.
+            """
+            # TODO (joao): This function should take an optional beam scorer function, to manipulate the scores after
+            # token selection. The function should be an argument exposed, so that custom scoring functions can be
+            # defined.
 
-                inputs_per_sub_batches = _split_model_inputs(
-                    model_inputs,
-                    split_size=batch_size,
-                    full_batch_size=batch_beam_size,
-                    config=self.config.get_text_config(),
+            # Gather the top K scores from _all_ beams.
+            if do_sample:
+                topk_indices = torch.multinomial(
+                    nn.functional.softmax(accumulated_log_probs, dim=-1), num_samples=beams_to_keep
                 )
-                outputs_per_sub_batch = [
-                    self(**inputs_per_sub_batch, return_dict=True) for inputs_per_sub_batch in inputs_per_sub_batches
-                ]
+                topk_log_probs = torch.gather(input=accumulated_log_probs, dim=1, index=topk_indices)
+            else:
+                topk_log_probs, topk_indices = torch.topk(accumulated_log_probs, k=beams_to_keep)
+
+            # Gather K top beams, recover the beam index by floor division and token id by modulo division
+            topk_current_beam_indices = topk_indices // vocab_size
+            topk_running_beam_indices = gather_beams(running_beam_indices, topk_current_beam_indices)
+            topk_running_sequences = gather_beams(running_sequences, topk_current_beam_indices)
+            topk_ids = topk_indices % vocab_size
+
+            # Update sequences for the K top-k new sequences.
+            topk_running_sequences[:, :, cur_len] = topk_ids
+
+            # we want to store the beam indices with batch information -> real beam index = beam index % num beams
+            batch_offset = torch.arange(batch_size, device=input_ids.device).view(-1, 1) * num_beams
+            batch_modified_indices = topk_current_beam_indices + batch_offset
+            topk_running_beam_indices[:, :, cur_len - decoder_prompt_len] = batch_modified_indices
+
+            return topk_log_probs, topk_running_sequences, topk_running_beam_indices
+
+        def get_running_beams_for_next_iteration(
```

Comment:
please place these functions just outside the context of beam search, they are unusable otherwise